### PR TITLE
🐛  Source Mixpanel - Add new datatime formats for state for cohort_members stream, added obsolete state reset for cohort_members stream

### PR DESCRIPTION
**Pull request by midavadim**
## What
ERROR # 1 cohorts stream
```
  File "/usr/local/lib/python3.9/site-packages/airbyte_cdk/sources/abstract_source.py", line 136, in read
    yield from self._read_stream(
  File "/usr/local/lib/python3.9/site-packages/airbyte_cdk/sources/abstract_source.py", line 236, in _read_stream
    for record_data_or_message in record_iterator:
  File "/usr/local/lib/python3.9/site-packages/airbyte_cdk/sources/streams/core.py", line 126, in read
    slices = self.stream_slices(
  File "/usr/local/lib/python3.9/site-packages/airbyte_cdk/sources/declarative/declarative_stream.py", line 142, in stream_slices
    return self.retriever.stream_slices()
  File "/usr/local/lib/python3.9/site-packages/airbyte_cdk/sources/declarative/retrievers/simple_retriever.py", line 375, in stream_slices
    return self.stream_slicer.stream_slices()
  File "/usr/local/lib/python3.9/site-packages/airbyte_cdk/sources/declarative/incremental/datetime_based_cursor.py", line 168, in stream_slices
    start_datetime = self._calculate_earliest_possible_value(self._select_best_end_datetime())
  File "/usr/local/lib/python3.9/site-packages/airbyte_cdk/sources/declarative/incremental/datetime_based_cursor.py", line 174, in _calculate_earliest_possible_value
    cursor_datetime = self._calculate_cursor_datetime_from_state(self.get_stream_state())
  File "/usr/local/lib/python3.9/site-packages/airbyte_cdk/sources/declarative/incremental/datetime_based_cursor.py", line 185, in _calculate_cursor_datetime_from_state
    return self.parse_date(stream_state[self._cursor_field.eval(self.config)])
  File "/usr/local/lib/python3.9/site-packages/airbyte_cdk/sources/declarative/incremental/datetime_based_cursor.py", line 234, in parse_date
    raise ValueError(f"No format in {self.cursor_datetime_formats} matching {date}")
ValueError: No format in ['%Y-%m-%d %H:%M:%S', '%Y-%m-%dT%H:%M:%S%z'] matching 2024-04-16T14:32:42
```

Error 2 - cohort_members stream
```
  File "/usr/local/lib/python3.9/site-packages/airbyte_cdk/sources/abstract_source.py", line 136, in read
    yield from self._read_stream(
  File "/usr/local/lib/python3.9/site-packages/airbyte_cdk/sources/abstract_source.py", line 236, in _read_stream
    for record_data_or_message in record_iterator:
  File "/usr/local/lib/python3.9/site-packages/airbyte_cdk/sources/streams/core.py", line 145, in read
    for record_data_or_message in records:
  File "/usr/local/lib/python3.9/site-packages/airbyte_cdk/sources/declarative/declarative_stream.py", line 120, in read_records
    yield from self.retriever.read_records(self.get_json_schema(), stream_slice)
  File "/usr/local/lib/python3.9/site-packages/airbyte_cdk/sources/declarative/retrievers/simple_retriever.py", line 324, in read_records
    for stream_data in self._read_pages(record_generator, self.state, _slice):
  File "/usr/local/lib/python3.9/site-packages/airbyte_cdk/sources/declarative/retrievers/simple_retriever.py", line 294, in _read_pages
    next_page_token = self._next_page_token(response)
  File "/usr/local/lib/python3.9/site-packages/airbyte_cdk/sources/declarative/retrievers/simple_retriever.py", line 258, in _next_page_token
    return self._paginator.next_page_token(response, self._records_from_last_response)
  File "/usr/local/lib/python3.9/site-packages/airbyte_cdk/sources/declarative/requesters/paginators/default_paginator.py", line 105, in next_page_token
    self._token = self.pagination_strategy.next_page_token(response, last_records)
  File "/airbyte/integration_code/source_mixpanel/components.py", line 278, in next_page_token
    if self._total and page_number is not None and self._total > self.page_size * (page_number + 1):
AttributeError: 'EngagePaginationStrategy' object has no attribute '_total'
```
Error 3 cohort_members::
```

		
  File "/usr/local/lib/python3.9/site-packages/airbyte_cdk/sources/abstract_source.py", line 136, in read
    yield from self._read_stream(
  File "/usr/local/lib/python3.9/site-packages/airbyte_cdk/sources/abstract_source.py", line 222, in _read_stream
    stream_instance.state = stream_state  # type: ignore # we check that state in the dir(stream_instance)
  File "/usr/local/lib/python3.9/site-packages/airbyte_cdk/sources/declarative/declarative_stream.py", line 85, in state
    self.retriever.state = state
  File "/usr/local/lib/python3.9/site-packages/airbyte_cdk/sources/declarative/retrievers/simple_retriever.py", line 385, in state
    self.cursor.set_initial_state(value)
  File "/usr/local/lib/python3.9/site-packages/airbyte_cdk/sources/declarative/incremental/per_partition_cursor.py", line 86, in set_initial_state
    for state in stream_state["states"]:
KeyError: 'states'
  },
```

Problem 4:
Fixing non-unique primary key for cohort_members stream
https://github.com/airbytehq/airbyte/issues/37833
Old key:  distinct_id
New key: distinct_id, cohort_id

## How
1. Added new datetime formats for state for cohort_members stream, 
2.  state for cohort_members stream is changed to new per-partition format
3. initialize total
4. changed key to distinct_id, cohort_id

## User Impact
Breaking change for `CohortMembers` stream: 
- State is changed to new empty per-partition format.
- Key is changed to new unique key (based on 'distinct_id' and 'cohort_id' fields) since previous key was not unique and didn't support possibility for user be in a few different cohorts.

## Can this PR be safely reverted and rolled back?

- [ ] YES 💚
- [x] NO ❌

### CODE REVIEW

File: airbyte-integrations/connectors/source-mixpanel/pyproject.toml [hunk 1]
```diff
@@ -3,7 +3,7 @@ requires = [ "poetry-core>=1.0.0",]
 build-backend = "poetry.core.masonry.api"
 
 [tool.poetry]
-version = "2.3.0"
+version = "2.3.1"
```
Comment (girarda) [inline airbyte-integrations/connectors/source-mixpanel/pyproject.toml:6]: this needs to be a breaking change because we're resetting the connection state

File: airbyte-integrations/connectors/source-mixpanel/source_mixpanel/manifest.yaml [hunk 2]
```diff
@@ -203,13 +205,27 @@ definitions:
...
+      - type: CustomStateMigration
+        class_name: "source_mixpanel.components.CohortMembersLegacyToPerPartitionStateMigration"
     retriever:
       $ref: "#/definitions/retriever"
       requester:
         $ref: "#/definitions/requester"
         http_method: POST
       paginator:
         $ref: "#/definitions/paginator"
+      record_selector:
```
Comment (girarda) [inline airbyte-integrations/connectors/source-mixpanel/source_mixpanel/manifest.yaml:218]: I'm very confused by this. why can't we use a normal incrmental_sync component?

**Review by girarda — CHANGES_REQUESTED** [review]

File: airbyte-integrations/connectors/source-mixpanel/pyproject.toml (hunk 1 above)
Comment (midavadim) [reply airbyte-integrations/connectors/source-mixpanel/pyproject.toml]: I have updated it to 3.0.0 now

File: airbyte-integrations/connectors/source-mixpanel/source_mixpanel/manifest.yaml (hunk 2 above)
Comment (midavadim) [reply airbyte-integrations/connectors/source-mixpanel/source_mixpanel/manifest.yaml:218]: It is not possible to filter records on API side.  So each time it returns all records and we filter them on our side.
 This behavior was migrated from the original python version of this connector.

File: airbyte-integrations/connectors/source-mixpanel/pyproject.toml [hunk 3]
```diff
@@ -18,6 +18,7 @@ include = "source_mixpanel"
 [tool.poetry.dependencies]
 python = "^3.9,<3.12"
 airbyte-cdk = "0.80.0"
+pytz = "^2024.1"
```
Comment (lazebnyi) [inline airbyte-integrations/connectors/source-mixpanel/pyproject.toml:21]: For what you need pytz?

File: airbyte-integrations/connectors/source-mixpanel/source_mixpanel/components.py [hunk 4]
```diff
@@ -337,3 +342,31 @@ def get_json_schema(self) -> Mapping[str, Any]:
...
+        self._config = config
+        self._parameters = parameters
+
+    def should_migrate(self, stream_state: Mapping[str, Any]) -> bool:
+        if _is_already_migrated(stream_state):
+            return False
+        return True
+
+    def migrate(self, stream_state: Mapping[str, Any]) -> Mapping[str, Any]:
+        return {"states": []}
```
Comment (lazebnyi) [inline airbyte-integrations/connectors/source-mixpanel/source_mixpanel/components.py:372]: Why we need reset state?

**Review by lazebnyi — CHANGES_REQUESTED** [review]

File: airbyte-integrations/connectors/source-mixpanel/pyproject.toml (hunk 3 above)
Comment (midavadim) [reply airbyte-integrations/connectors/source-mixpanel/pyproject.toml]: it is removed

File: airbyte-integrations/connectors/source-mixpanel/source_mixpanel/components.py (hunk 4 above)
Comment (midavadim) [reply airbyte-integrations/connectors/source-mixpanel/source_mixpanel/components.py:372]: Do you mean that state reset is not needed because after connector update  data for cohort_members stream will be reset?

Comment (midavadim) [reply airbyte-integrations/connectors/source-mixpanel/source_mixpanel/components.py:372]: I have removed custom state migration

**Comment by midavadim** [comment]
Regression test:
https://github.com/airbytehq/airbyte/actions/runs/9082610503

File: airbyte-integrations/connectors/source-mixpanel/source_mixpanel/components.py [hunk 5]
```diff
@@ -9,7 +9,10 @@
 from airbyte_cdk.models import AirbyteMessage, SyncMode, Type
 from airbyte_cdk.sources.declarative.extractors import DpathExtractor
 from airbyte_cdk.sources.declarative.interpolation import InterpolatedString
-from airbyte_cdk.sources.declarative.migrations.legacy_to_per_partition_state_migration import LegacyToPerPartitionStateMigration
+from airbyte_cdk.sources.declarative.migrations.legacy_to_per_partition_state_migration import (
+    LegacyToPerPartitionStateMigration,
+    _is_already_migrated,
```
Comment (lazebnyi) [inline airbyte-integrations/connectors/source-mixpanel/source_mixpanel/components.py:14]: Why do we need `_is_already_migrated` ?

File: airbyte-integrations/connectors/source-mixpanel/metadata.yaml [hunk 6]
```diff
@@ -33,6 +33,13 @@ data:
   releaseStage: generally_available
   releases:
     breakingChanges:
+      3.0.0:
+        message:
+          In this release, state for CohortMembers is changed to per partition format.
+          Key for CohortMembers stream was changed to new unique key based on 
+          distinct_id and cohort_id fields since previous key was not unique and 
+          didn't support possibility for user be in a few different cohorts
+        upgradeDeadline: "2024-05-22"
```
Comment (lazebnyi) [inline airbyte-integrations/connectors/source-mixpanel/metadata.yaml:42]: please set upgradeDeadline to 2024-06-03

**Review by lazebnyi — CHANGES_REQUESTED** [review]

File: airbyte-integrations/connectors/source-mixpanel/source_mixpanel/components.py (hunk 5 above)
Comment (midavadim) [reply airbyte-integrations/connectors/source-mixpanel/source_mixpanel/components.py]: removed

File: airbyte-integrations/connectors/source-mixpanel/metadata.yaml (hunk 6 above)
Comment (midavadim) [reply airbyte-integrations/connectors/source-mixpanel/metadata.yaml]: updated

File: docs/integrations/sources/mixpanel-migrations.md [hunk 7]
```diff
@@ -1,5 +1,63 @@
...
+2. Select the instance of the connector you wish to upgrade.
+
+:::note
+Each instance of the connector must be updated separately. If you have created multiple instances of a connector, updating one will not affect the others.
+:::
+
+3. Select **Upgrade**
+    1. Follow the prompt to confirm you are ready to upgrade to the new version.
+
+### Refresh affected schemas and reset data
```
Comment (lazebnyi) [inline docs/integrations/sources/mixpanel-migrations.md:39]: Do breaking changes affect the schema?

File: airbyte-integrations/connectors/source-mixpanel/unit_tests/test_streams.py [hunk 8]
```diff
@@ -171,17 +171,71 @@ def test_engage_stream_incremental(requests_mock, engage_response, config_raw):
...
+              "id": 1111,
+              "parent_slice": {}
+            },
+            "cursor": {
+              "last_seen": "2024-02-01T11:20:47"
+            }
+        }]
+    }
+
+    # stream_state = {"states":[ {"last_seen": "2024-02-01T11:20:47"}] }
```
Comment (lazebnyi) [inline airbyte-integrations/connectors/source-mixpanel/unit_tests/test_streams.py:226]: dead code

**Review by lazebnyi — CHANGES_REQUESTED** [review]

File: docs/integrations/sources/mixpanel-migrations.md (hunk 7 above)
Comment (midavadim) [reply docs/integrations/sources/mixpanel-migrations.md:39]: In order to update primary key for cohort members client should click on "Refresh Soutce Schema"

File: airbyte-integrations/connectors/source-mixpanel/unit_tests/test_streams.py (hunk 8 above)
Comment (midavadim) [reply airbyte-integrations/connectors/source-mixpanel/unit_tests/test_streams.py:226]: removed

**Review by lazebnyi — CHANGES_REQUESTED** [review]
I have double-checked those regression tests that you ran.

And I found that you did not test cohorts at all. Also, you need to check the engagement stream and request quantity, as tests failed for them here. - https://ff8e0fd38e62a763fed2b06b1783398be160d5874d6a2caa0e1ec6d-apidata.googleusercontent.com/download/storage/v1/b/airbyte-ci-reports-multi/o/airbyte-ci%2Fconnectors%2Ftest%2Fmanual%2Fmidavadim_mixpanel-datatime-fix%2F1715768566%2Fdd88ee7c3f0e50b1b94f86375d8ac7c775e01cf9%2Fsource-mixpanel%2F3.0.0%2Foutput.html?jk=ASOpP9h2FRxXPWvb_gAFPyCpnKRtEulBQaKJUDwWzlz4y-8XmQrEt4nsVP5tmbYEbNYHhzGt96FJ0ui1dO2JZjV4SV7yDdnkM_oWJSa1VdRCQO5_R6LFTnHLKaV72gUMUFb-an3utbQi5BqJfwI-PUIOPD2wnJxD-Oog1OOV9OpU1D4SNtuyn5QFzeTHlYyQmpEF4FtJbwjyBiizwaAS5j8j-sFF8r1BXdOzxkOXeKlgxbBX4JXna6Id53Tn1IHPhGsHWdrHCiOgdvCy4ElOv_ArfrSENjctPv9oWkWocqfIo1blzANNiO9SmICTLxImSw9VEbaJhAa-hy16UzXyAmOzW9zt-Te0N5uGoN-7BzceRGsy0oWGTs6MQq5p1UXTUIfU83Gf-TcUOtlMBFfiKjkFeO3JBpFDUIQPJVa1wTakyc3Fch58cPePJ8fq0xl6wb0PtOyRTy67lBI4eCNKY5B46EthooVrc6TZ30k2eRF2lxbZUttoZpzSaC6hsrL2q7xi8PppK5u7GSSBtOtS_xpjVG_ufvx30LBacgknqjYyeY3DlQe2yxZdGZd6L-OWRKzH-pzzbZGPfqWN2uC3g06QrukXw0NgyAaBhqmqMjidFqjOhnCbtwpKA3pd4XHENXo9X8UeiHsntRs5Cw3uB6-0gmQ-TUDkMWxAh1pq_znOrTKAJKuNw3BI9oAZsNCQmI15zdknzFH1Lcgi2qmFR2fRWsiXSjqcvR-3wNGGnw1S1WaX3d_TfYcBwhgo7i2uKn0RtXam0QBSdzqgb-DKUAvXAWduGzrIzXz8VIgtq2eiyHARRppg_D_rZEzDcL5msNY0R0509_0VxsO9JtfXEv6xO8Pk2nMVvDO_fUUJeT13MDB-1UrEZlhpYGFlSilUKGK2GV8-PPST-oKhXIDsn4wMqXR-2lod0NE66Mni5Vt4rz3Tk6GmymW2JRJUoD8vZDUAomzNPGOaeTk56DP8SDfAe-fr28p6ZnVD-6rWVa6xniiu4yRTMAw1IvrxdjGoHbbV1UjBvepZLQbD-hs7Pl8brwFbDf4tt0Z-aIeAhzigrQqsnBjdkjOqS5N_TsVJ-GhLi9CKPKdHnUoK1oOrpjryWexQSV5JAU0gTON06F9k8zQXBxmc3bggorANGzUAvRnf_KpZ3xXFQiwp_x-rAAc9Kg1x0hjWVVC1KrWc_O9h0E7KAYHWyt3rQnUJ-wwC_vm4eEXWSW1yNchWjZWJ8ASr0PEQ4umyDUzHnkGxUoOWmwr6YPTcjjXzI7CeQFqIwQUEuYBcQ1-7_KTW4XacRJHGaftpOA1_E8ZUnmP2RYFnHC7DwieSk4wd0FwBj5Xr8Fq003koc1AhA6JuXdLjZd85sBSVv-HS41BDopZCN323-DFqAS794Mm6-pgNxFE7_JtdvQhzr2_F_oK5i5KsjeXX7A&isca=1

**Comment by midavadim** [comment]
> I have double-checked those regression tests that you ran.
> 
> And I found that you did not test cohorts at all

1. As I said before  cohorts and cohort_members are semi incremental streams so during each sync it reads all data and cover 95% of regular incremental. Of course Incremental with data in better.

2. So I have tested incremental syncs for cohorts and cohort_members stream with custom configured_catalog and state.
The current connector version works fine, **but the number of records is different** because old  client-side filtering did not work correctly on non-chronological data.


Comments for Regression tool report:
 1. "Untested streams (not in configured catalog or without data)":  There is a difference in when stream was not configured to run or it was run but returned 0 records (which is very common for incremental streams)
 2. OVERALL: Report does not indicate errors that happened during sync. They can be only found in command_execution_artifacts/source-mixpanel/read-with-state/dev/stdout.log.  Now report says that it returns 0 records (see item 1). There is a difference in when stream read 0 records (because it is an incremental stream with up-to-date state) or failed during state initialization or due to API rate limits or another problems.
 3. "Requested URLs": Probably it would be good to add HTTP body (if it is specified) to URLs as well. Otherwise we need to look at raw data (http_dump.har) to understand why there are the same URLs - is it because of retries or has different bodies

**Review by lazebnyi — APPROVED** [review]
LGTM

File: airbyte-integrations/connectors/source-mixpanel/source_mixpanel/manifest.yaml [hunk 9]
```diff
@@ -129,12 +129,16 @@ definitions:
       record_selector:
         $ref: "#/definitions/selector_empty_dpath"
         record_filter:
-          condition: "{{ record['created'] >= stream_state.created if stream_state.created else true }}"
+          condition: |
+            {%- set filter_date = stream_state.created or stream_slice.start_time -%}
+            "{{ record['created'] >= filter_date if filter_date else true }}"
```
Comment (girarda) [inline airbyte-integrations/connectors/source-mixpanel/source_mixpanel/manifest.yaml:134]: when would `filter_date` not be true?

I'd expect us to be able to simplify the filter to
`{{ record['created'] >= stream_interval["start_time"] }}`

Mid-term, we should stop exposing the state to connector developer because it's often misused and leads to bugs that are often difficult to troubleshoot. [Issue for reference](https://github.com/airbytehq/airbyte-internal-issues/issues/2209)

File: airbyte-integrations/connectors/source-mixpanel/source_mixpanel/components.py [hunk 10]
```diff
@@ -261,6 +259,8 @@ class EngagePaginationStrategy(PageIncrement):
     page - incremental page number
     """
 
+    _total = 0
```
Comment (girarda) [inline airbyte-integrations/connectors/source-mixpanel/source_mixpanel/components.py:262]: shouldn't this be reset from the `reset` method since this is part of the inner state?

File: airbyte-integrations/connectors/source-mixpanel/source_mixpanel/manifest.yaml [hunk 11]
```diff
@@ -232,7 +252,7 @@ definitions:
         fields:
           - path:
               - browser_version
-            value: "{{ record.browser_version | string }}"
+            value: "{{ record.browser_version | string if record.browser_version else '' }}"
```
Comment (girarda) [inline airbyte-integrations/connectors/source-mixpanel/source_mixpanel/manifest.yaml:255]: did we see a case where this happens? It's not clear why this change is required from the PR description

File: airbyte-integrations/connectors/source-mixpanel/source_mixpanel/manifest.yaml [hunk 12]
```diff
@@ -210,6 +219,17 @@ definitions:
         http_method: POST
       paginator:
         $ref: "#/definitions/paginator"
+      record_selector:
+        $ref: "#/definitions/selector"
+        record_filter:
+          condition: |
+            {%- set partition_state = namespace(value=config.start_date) -%}
```
Comment (girarda) [inline airbyte-integrations/connectors/source-mixpanel/source_mixpanel/manifest.yaml:226]: @artem1205 FYI - flagging that we need to make sure the our semi-incremental implementation works for substreams.

Comment (girarda) [reply airbyte-integrations/connectors/source-mixpanel/source_mixpanel/manifest.yaml:226]: @midavadim is support for semi-incremental required for the bug fix or could this be released in 3.0.1? reason is that `3.0.0` will not be revertible so I'd prefer keeping the scope of the change as small as possible

File: airbyte-integrations/connectors/source-mixpanel/source_mixpanel/manifest.yaml [hunk 13]
```diff
@@ -210,6 +219,17 @@ definitions:
...
+            {%- for state in stream_state.get('states', []) -%}
```
Comment (girarda) [inline airbyte-integrations/connectors/source-mixpanel/source_mixpanel/manifest.yaml:227]: @midavadim this is a pretty complicated condition. can you add tests to cover all branches?

**Review by girarda — CHANGES_REQUESTED** [review]
left a few comments. main questions
- can semi-incremental be released separately from the bug fixes?
- can the filters only rely on stream_interval instead of stream_state?

File: airbyte-integrations/connectors/source-mixpanel/source_mixpanel/manifest.yaml (hunk 13 above)
Comment (midavadim) [reply airbyte-integrations/connectors/source-mixpanel/source_mixpanel/manifest.yaml]: I have added unittests for this case

File: airbyte-integrations/connectors/source-mixpanel/source_mixpanel/manifest.yaml (hunk 11 above)
Comment (midavadim) [reply airbyte-integrations/connectors/source-mixpanel/source_mixpanel/manifest.yaml:244]: 1. If we do not add conversion to string then I see records where value can be a number on our account.
2. During tests I also saw that when 'browser_version' is missing in records then value in record:
"browser_version": "{{ record.browser_version11111 | string if record.browser_version else '' }}"

So this update is needed

File: airbyte-integrations/connectors/source-mixpanel/source_mixpanel/manifest.yaml (hunk 9 above)
Comment (midavadim) [reply airbyte-integrations/connectors/source-mixpanel/source_mixpanel/manifest.yaml]: I have rewritten this part with stream_interval usage

**Comment by midavadim** [comment]
> left a few comments. main questions
> 
> * can semi-incremental be released separately from the bug fixes?
> * can the filters only rely on stream_interval instead of stream_state?

1. "semi-incremental" - WIthout filtering each incremental syncs returns all records. it is implemented in order to support proper 'incremental' sync which existed in python version of this connector (2.2.0). When we were discussing this change, it was agreed that we must implement the same behavior as it was before low code otherwise it is treated as another breaking change. Please correct me if this is a wrong statement.

3. "stream_interval " - I have rewritten filters to use stream_interval . It is indeed much more comfortable to use instead of stream_state and stream_slice, especially for streams with partitioning.

File: airbyte-integrations/connectors/source-mixpanel/source_mixpanel/manifest.yaml (hunk 12 above)
Comment (midavadim) [reply airbyte-integrations/connectors/source-mixpanel/source_mixpanel/manifest.yaml]: @girarda  "semi-incremental" - WIthout filtering each incremental syncs returns all records. it is implemented in order to support proper 'incremental' sync which existed in python version of this connector (2.2.0). When we were discussing this change, it was agreed that we must implement the same behavior as it was before low code otherwise it is treated as another breaking change. Please correct me if this is a wrong statement.

File: airbyte-integrations/connectors/source-mixpanel/source_mixpanel/components.py (hunk 10 above)
Comment (midavadim) [reply airbyte-integrations/connectors/source-mixpanel/source_mixpanel/components.py:262]: I have added reset function. But it is not directly required

File: docs/integrations/sources/mixpanel.md [hunk 14]
```diff
@@ -54,42 +54,43 @@ Syncing huge date windows may take longer due to Mixpanel's low API rate-limits
...
-| 0.1.20 | 2022-08-22 | [15091](https://github.com/airbytehq/airbyte/pull/15091) | Improve `export` stream cursor support |
-| 0.1.19 | 2022-08-18 | [15739](https://github.com/airbytehq/airbyte/pull/15739) | Update `titile` and `description` for `Project Secret` field |
-| 0.1.18 | 2022-07-21 | [14924](https://github.com/airbytehq/airbyte/pull/14924) | Remove `additionalProperties` field from schemas and specs |
+|:--------| :--------- | :------------------------------------------------------- | :---------------------------------------------------------------------------------------------------------- |
+| 3.0.0   | 2024-05-10 | [38066](https://github.com/airbytehq/airbyte/pull/38066) | Changed key to distinct_id, cohort_id and changed state to per-patition format for `CohortMembers` stream; fixed pagination for `Engage` stream; fixed incorrect client-side filtering for semi-incremental streams when data comes not in chronological order; semi-incremental `Cohorts`, `CohortMembers` and `Engage` streams with client-side filtering extract records since user provided or default (1 year old) start_date |
```
Comment (girarda) [inline docs/integrations/sources/mixpanel.md:58]:
```suggestion
| 3.0.0   | 2024-05-21 | [38066](https://github.com/airbytehq/airbyte/pull/38066) | Changed key to distinct_id, cohort_id and changed state to per-patition format for `CohortMembers` stream; fixed pagination for `Engage` stream; fixed incorrect client-side filtering for semi-incremental streams when data comes not in chronological order; semi-incremental `Cohorts`, `CohortMembers` and `Engage` streams with client-side filtering extract records since user provided or default (1 year old) start_date |
```

**Review by girarda — APPROVED** [review]
looks good thanks @midavadim !
:shipit:

**Review by katmarkham — CHANGES_REQUESTED** [review]

File: airbyte-integrations/connectors/source-mixpanel/metadata.yaml [hunk 15]
```diff
@@ -26,18 +26,34 @@ data:
...
-      dockerImageTag: 2.2.0 # temporary pin due to a bug in 2.3.0 https://github.com/airbytehq/airbyte/pull/38106
     oss:
       enabled: true
-      dockerImageTag: 2.2.0 # temporary pin due to a bug in 2.3.0 https://github.com/airbytehq/airbyte/pull/38106
   releaseStage: generally_available
   releases:
     breakingChanges:
+      3.0.0:
+        message:
+          In this release, state for CohortMembers is changed to per partition format.
```
[the remark this answers is not in the document]
Comment (lazebnyi) [reply airbyte-integrations/connectors/source-mixpanel/metadata.yaml]: Updated

File: airbyte-integrations/connectors/source-mixpanel/metadata.yaml [hunk 16]
```diff
@@ -26,18 +26,30 @@ data:
...
-      dockerImageTag: 2.2.0 # temporary pin due to a bug in 2.3.0 https://github.com/airbytehq/airbyte/pull/38106
     oss:
       enabled: true
-      dockerImageTag: 2.2.0 # temporary pin due to a bug in 2.3.0 https://github.com/airbytehq/airbyte/pull/38106
   releaseStage: generally_available
   releases:
     breakingChanges:
+      3.0.0:
+        message:
+          In this release, CohortMembers stream has changed for wrong pk and state format.
```
Comment (katmarkham) [inline airbyte-integrations/connectors/source-mixpanel/metadata.yaml:36]:
```suggestion
          In this release, CohortMembers stream has changed due to changes in primary key and an improper state format. 
```

**Review by katmarkham — APPROVED** [review]
